### PR TITLE
Mrtk version next input action extensions / updates

### PR DESCRIPTION
**Pull request by SimonDarksideJ**
Overview
---
Updates from the device branch to improve the extensibility of the InputActions system

Changes
---
- Resolved namespaces for files in the Internal.Definitions folders
- Added AxisType to InputAction for easier filtering on valid axis
- Added InputHoldAction property to InteractionDefinition (still needs updating in custom editor)
- Changes InputSource.Interactions to use a Dictionary instead of an Array for easier lookup by devices
- Added default loading routines to InputActionsProfile for use by devices with no pre-setup config

### CODE REVIEW

File: Assets/MixedRealityToolkit/InputSystem/Sources/BaseGenericInputSource.cs [hunk 1]
```diff
@@ -1,10 +1,11 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-using Microsoft.MixedReality.Toolkit.Internal.Definitions.InputSystem;
+using Microsoft.MixedReality.Toolkit.Internal.Definitions;
 using Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem;
 using Microsoft.MixedReality.Toolkit.Internal.Managers;
 using System.Collections;
+using System.Collections.Generic;
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/InputSystem/Sources/BaseGenericInputSource.cs]: why include in some source files and not others that use Dictionary?

File: Assets/MixedRealityToolkit/InputSystem/Sources/DictationInputSource.cs [hunk 2]
```diff
@@ -1,11 +1,11 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-using Microsoft.MixedReality.Toolkit.Internal.Definitions.InputSystem;
 using Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem;
 using Microsoft.MixedReality.Toolkit.Internal.Utilities.Async;
 using System.Threading.Tasks;
 using UnityEngine;
+using Microsoft.MixedReality.Toolkit.Internal.Definitions;
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/InputSystem/Sources/DictationInputSource.cs]: nit: not in alphabetical order

File: Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputActionsProfile.cs [hunk 3]
```diff
@@ -11,6 +12,9 @@ namespace Microsoft.MixedReality.Toolkit.Internal.Definitions.InputSystem
     [CreateAssetMenu(menuName = "Mixed Reality Toolkit/Mixed Reality Input Actions Profile", fileName = "MixedRealityInputActionsProfile", order = 1)]
     public class MixedRealityInputActionsProfile : ScriptableObject
     {
+        private readonly string[] inputActionsMRTKDefaults = { "Select", "Menu", "Grip", "Pointer", "Walk", "Look", "Interact", "Pickup", "Inventory", "ConversationSelect" };
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputActionsProfile.cs]: would be a bit easier to read if each element were on a separate line

File: Assets/MixedRealityToolkit/_Core/Definitions/InteractionDefinition.cs [hunk 4]
```diff
@@ -41,6 +59,10 @@ public InteractionDefinition(uint id, AxisType axisType, Devices.DeviceInputType
         /// </summary>
         public InputAction InputAction { get; set; }
 
+        /// <summary>
+        /// Action to be raised to the Input Manager when the input data has changed.
+        /// </summary>
+        public InputAction InputHoldAction { get; set; }
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/InteractionDefinition.cs]: Should we also include a Pointer action as well?

Originally I had these defined in the Input Action Profile itself.

Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Definitions/InteractionDefinition.cs]: Except that Pointer is unique to pointers ONLY, Hold is a held state for any input

File: Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputActionsProfile.cs (hunk 3 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputActionsProfile.cs]: #Grumble, only a default :P

File: Assets/MixedRealityToolkit/InputSystem/Sources/BaseGenericInputSource.cs (hunk 1 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/InputSystem/Sources/BaseGenericInputSource.cs]: Please elaborate @davidkline-ms ?

Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/InputSystem/Sources/BaseGenericInputSource.cs]: Ok, sorted out the issue I believe you were pointing out.

File: Assets/MixedRealityToolkit/_Core/Definitions/InteractionDefinition.cs (hunk 4 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/InteractionDefinition.cs]: not true, it was a global action for all pointers

File: Assets/MixedRealityToolkit/InputSystem/Sources/TouchscreenInputSource.cs [hunk 5]
```diff
@@ -38,13 +38,13 @@ public TouchPointer(string name, Touch touch, Ray ray, IMixedRealityInputSource
...
-            : base("Touch Input", new[]
+            : base("Touch Input", new Dictionary<Internal.Definitions.Devices.DeviceInputType, InteractionDefinition>()
         {
-            new InteractionDefinition(1, AxisType.DualAxis, Internal.Definitions.Devices.DeviceInputType.Touchpad),
-            new InteractionDefinition(2, AxisType.DualAxis, Internal.Definitions.Devices.DeviceInputType.PointerPosition),
-            new InteractionDefinition(3, AxisType.Digital, Internal.Definitions.Devices.DeviceInputType.TouchpadTouch),
-            new InteractionDefinition(4, AxisType.Digital, Internal.Definitions.Devices.DeviceInputType.TouchpadPress),
-            new InteractionDefinition(5, AxisType.Digital, Internal.Definitions.Devices.DeviceInputType.PointerClick),
+            {Internal.Definitions.Devices.DeviceInputType.Touchpad, new InteractionDefinition(1, AxisType.DualAxis, Internal.Definitions.Devices.DeviceInputType.Touchpad) },
+            {Internal.Definitions.Devices.DeviceInputType.PointerPosition, new InteractionDefinition(2, AxisType.DualAxis, Internal.Definitions.Devices.DeviceInputType.PointerPosition) },
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/InputSystem/Sources/TouchscreenInputSource.cs]: I don't think using the device input type as the key is the right thing to do.

File: Assets/MixedRealityToolkit/_Core/Definitions/InteractionDefinition.cs [hunk 6]
```diff
@@ -19,6 +20,23 @@ public InteractionDefinition(uint id, AxisType axisType, Devices.DeviceInputType
...
+            InputAction = inputAction;
+        }
+
+        public InteractionDefinition(uint id, AxisType axisType, Devices.DeviceInputType inputType, InputAction inputAction, InputAction inputHoldAction) : this()
+        {
+            Id = id;
+            AxisType = axisType;
+            InputType = inputType;
+            InputAction = inputAction;
+            InputHoldAction = inputHoldAction;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/InteractionDefinition.cs]: What about pointer actions?

File: Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputActionsProfile.cs [hunk 7]
```diff
@@ -11,6 +12,34 @@ namespace Microsoft.MixedReality.Toolkit.Internal.Definitions.InputSystem
...
+        {
+            "Select",
+            "Menu",
+            "Grip",
+            "Pointer",
+            "Walk",
+            "Look",
+            "Interact",
+            "Pickup",
+            "Inventory",
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputActionsProfile.cs]: A few of these I'm not sure fits all use cases.

I thin this should def be something we leave open for devs to setup on their own.

File: Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputActionsProfile.cs [hunk 8]
```diff
@@ -11,6 +12,34 @@ namespace Microsoft.MixedReality.Toolkit.Internal.Definitions.InputSystem
...
+            "ConversationSelect"
+        }; // Examples only, to be refined later.
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputActionsProfile.cs]: I know these are examples, but should we keep them limited to the basic ones, like Select, Menu, Grip, Pointer, Teleport, etc?

File: Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputActionsProfile.cs [hunk 9]
```diff
@@ -34,5 +63,42 @@ public class MixedRealityInputActionsProfile : ScriptableObject
...
+        {
+            List<InputAction> defaultMRTKActions = new List<InputAction>();
+            bool exists = false;
+
+            for (uint i = 0; i < inputActionsMRTKDefaults.Length; i++)
+            {
+                defaultMRTKActions.Add(new InputAction(i, inputActionsMRTKDefaults[i], inputActionsMRTKDefaultsAxis[i]));
+            }
+
+            foreach (var currentAction in inputActions)
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/InputSystem/MixedRealityInputActionsProfile.cs]: we never use the currentAction?

**Comment by SimonDarksideJ** [comment]
Reviewing after late night brainstorming

**Pull request closed without merging**